src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java

```diff
@@ -0,0 +1,389 @@
+                client.index(request, ActionListener.runBefore(listener, () -> context.restore()));
+            } catch (Exception e) {
+                logger.error("Failed to update global_context entry : {}. {}", documentId, e.getMessage());
+                listener.onFailure(e);
+            }
+        }
+    }
+
+    /**
+     * Updates a document in the workflow state index
+     * @param workflowStateDocId the document ID
+     * @param updatedFields the fields to update the global state index with
+     * @param listener action listener
+     */
+    public void updateWorkflowState(String workflowStateDocId, Map<String, Object> updatedFields, ActionListener<UpdateResponse> listener) {
+        if (!doesIndexExist(WORKFLOW_STATE_INDEX)) {
+            String exceptionMessage = "Failed to update state for given workflow due to missing workflow_state index";
+            logger.error(exceptionMessage);
+            listener.onFailure(new Exception(exceptionMessage));
+        } else {
+            try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+                UpdateRequest updateRequest = new UpdateRequest(WORKFLOW_STATE_INDEX, workflowStateDocId);
+                Map<String, Object> updatedContent = new HashMap<>();
+                updatedContent.putAll(updatedFields);
+                updateRequest.doc(updatedContent);
+                updateRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+                client.update(updateRequest, ActionListener.runBefore(listener, () -> context.restore()));
+            } catch (Exception e) {
+                logger.error("Failed to update workflow_state entry : {}. {}", workflowStateDocId, e.getMessage());
+                listener.onFailure(e);
+            }
+        }
+    }
+
+    /**
+     * Update global context index for specific fields
+     * @param documentId global context index document id
+     * @param updatedFields updated fields; key: field name, value: new value
+     * @param listener UpdateResponse action listener
+     */
+    public void storeResponseToGlobalContext(
+        String documentId,
+        Map<String, Object> updatedFields,
+        ActionListener<UpdateResponse> listener
+    ) {
+        UpdateRequest updateRequest = new UpdateRequest(GLOBAL_CONTEXT_INDEX, documentId);
+        Map<String, Object> updatedUserOutputsContext = new HashMap<>();
+        updatedUserOutputsContext.putAll(updatedFields);
+        updateRequest.doc(updatedUserOutputsContext);
+        updateRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+        // TODO: decide what condition can be considered as an update conflict and add retry strategy
+
+        try {
+            client.update(updateRequest, listener);
+        } catch (Exception e) {
+            logger.error("Failed to update global_context index");
+            listener.onFailure(e);
+        }
```

Comment:
`updateWorkflowState` and `storeResponseToGlobalContext` methods can be combined together. Since both are updating the document in the index.